### PR TITLE
error handling when path to dir provided instead of file

### LINKED ISSUE #109450: Flutter tools crashes when path for `flutter screenshot` is specified as `.`
## command

flutter screenshot -o . -v

## exception

FileSystemException: FileSystemException: Cannot retrieve length of file, path = '.' (OS Error: The request is not supported.
, errno = 50)

```
#0      _File.throwIfError (dart:io/file_impl.dart:635:7)
#1      _File.lengthSync (dart:io/file_impl.dart:377:5)
#2      ForwardingFile.lengthSync (package:file/src/forwarding/forwarding_file.dart:35:32)
#3      ForwardingFile.lengthSync (package:file/src/forwarding/forwarding_file.dart:35:32)
#4      ScreenshotCommand._showOutputFileInfo (package:flutter_tools/src/commands/screenshot.dart:199:36)
#5      ScreenshotCommand.runScreenshot (package:flutter_tools/src/commands/screenshot.dart:135:5)
<asynchronous suspension>
#6      ScreenshotCommand.runCommand (package:flutter_tools/src/commands/screenshot.dart:110:9)
<asynchronous suspension>
#7      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1209:27)
<asynchronous suspension>
#8      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#9      CommandRunner.runCommand (package:args/command_runner.dart:209:13)
<asynchronous suspension>
#10     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:281:9)
<asynchronous suspension>
#11     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#12     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:229:5)
<asynchronous suspension>
#13     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:62:9)
<asynchronous suspension>
#14     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#15     main (package:flutter_tools/executable.dart:91:3)
<asynchronous suspension>
```

## flutter doctor

```
[✓] Flutter (Channel beta, 3.3.0-0.3.pre, on Microsoft Windows [Version 10.0.19043.1826], locale en-IN)
    • Flutter version 3.3.0-0.3.pre on channel beta at C:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 8c1149878b (2 days ago), 2022-08-10 12:15:36 -0500
    • Engine revision f16e757d5d
    • Dart version 2.18.0 (build 2.18.0-271.4.beta)
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\Users\moham\AppData\Local\Android\sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✗] Visual Studio - develop for Windows
    ✗ Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[✓] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2022.1.4
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.70.1)
    • VS Code at C:\Users\moham\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.46.0

[✓] Connected device (4 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows       • windows-x64    • Microsoft Windows [Version 10.0.19043.1826]
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 103.0.5060.134
    • Edge (web)              • edge          • web-javascript • Microsoft Edge 103.0.1264.44

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by eliasyishak**
Fixes: #109450

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by eliasyishak** [comment]
Need to create a test to be run in the device lab following instructions found here: https://github.com/flutter/flutter/blob/master/dev/devicelab/README.md

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 1]
```diff
@@ -123,16 +127,41 @@ class ScreenshotCommand extends FlutterCommand {
...
+        'Directory: "${outputFile.dirname}" does not exist'
+      );
+    }
+
+    // Conditional for validating filename with allowed extensions
+    if (!RegExp(_kFileNamePattern).hasMatch(outputFile.basename)) {
+      throwToolExit(
+        'The provided filename is invalid\n'
+        'Allowed filename characters: "a-z", "A-Z", "0-9", "_", "-"\n'
+        'Allowed extensions: jpeg, jpg, png)\n'
```
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/commands/screenshot.dart:152]: Might want to revisit this to determine what file types are supported through all targets

**Comment by jonahwilliams** [comment]
I don't think this is the right approach: in general this feels far too restrictive, especially as it is perfectly reasonable for folks to give screenshots whatever extension they want or use non-ascii characters.


Is there an exception thrown if we create a file with an unsupported name, like "."? If so, we could catch that exception instead. But consider shifting from an allow based approach to a deny based approach

**Comment by eliasyishak** [comment]
Yeah it definitely was a more restrictive approach, and that's understandable if we would like to make it less restrictive. We can remove any character restriction and file type restriction on the pattern and just make sure the directory is valid?

Because the tool does actually allow for files that don't even have any extensions, ie. `-o image` will produce a file for the user. Does that approach make sense?

**Comment by christopherfujino** [comment]
> Yeah it definitely was a more restrictive approach, and that's understandable if we would like to make it less restrictive. We can remove any character restriction and file type restriction on the pattern and just make sure the directory is valid?
> 
> Because the tool does actually allow for files that don't even have any extensions, ie. `-o image` will produce a file for the user. Does that approach make sense?

isn't it reading the extension and then determining what encoding to use based on that? so if you pass `-o image`, what kind of image is it creating? is there a default extension, or is it just not doing anything and exiting 0?

**Comment by jonahwilliams** [comment]
As far as I can tell, the tool does not use the file extension to determine what kind of file output.

**Comment by christopherfujino** [comment]
> As far as I can tell, the tool does not use the file extension to determine what kind of file output.

this is simctl

**Comment by eliasyishak** [comment]
I'm not sure how simctl handles it but I have seen that the file gets generated, although I wasn't able to open it on my mac. Which was why I initially thought to add the extension requirement and possibly expand to different file types as needed.

**Comment by jonahwilliams** [comment]
but simctl isn't the only backend here, and based on the snapshot type argument we can also create a "skia" and "rasterizer" type

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 2]
```diff
@@ -183,6 +191,37 @@ class ScreenshotCommand extends FlutterCommand {
     return true;
   }
 
+  /// Additional error checks below for valid file name
+  /// and directory because ios/xcode screenshot binary will
+  /// return a 0 exit code even if outputFile is not valid
+  static File validateOutputFile(File outputFile, FileSystem fs) {
+
+    // Will make sure that if the user passes '-o image.png/' the file
```
Comment (eliasyishak) [inline packages/flutter_tools/lib/src/commands/screenshot.dart:199]: @jonahwilliams i added this additional functionality to catch any errors if the user provides a trailing delimiter. Do you think it's worth keeping this or should the tool throw an exception?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/screenshot.dart:199]: I would not modify the file the user provides to this script, either accept it if it is valid or deny if it isn't. Consider the case where someone uses this command in part of an automated test script, and said test script has a bug that gets an invalid path. You would want that to fail as soon as possible, instead of accidentally passing for some time, or failing because a different command that accepts a file didn't make the same change to the file name to make it valid

Comment (eliasyishak) [reply packages/flutter_tools/lib/src/commands/screenshot.dart:199]: Great point, I'll make the adjustments

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 3]
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
...
 
+  /// Additional error checks below for valid file name
+  /// and directory because ios/xcode screenshot binary will
+  /// return a 0 exit code even if outputFile is not valid
+  static void validateOutputFile(File outputFile, FileSystem fs) {
+
+    const Platform platform = LocalPlatform();
+
+    // Conditional to check for trailing path separator
+    if (outputFile.path.endsWith(platform.pathSeparator)) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/screenshot.dart:202]: This is available on `path`.

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 4]
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
...
+      throwToolExit(
+        'The provided path cannot end with a path separator\n'
+        'Path provided: "${outputFile.path}"'
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/screenshot.dart:205]: I would rephrase this to indicate that we expect a file path but got a directory path.

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 5]
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
...
+      );
+    }
+
+    // Conditional for validating directory is valid
+    if (!fs.directory(outputFile.dirname).existsSync()) {
+      throwToolExit(
+        'The provided path to file needs to have a directory that exists\n'
+        'Directory: "${outputFile.dirname}" does not exist'
+      );
+    }
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/screenshot.dart:215]: Use `fs.file(outputFile).parent.existsSync()` - I'm not sure if dirname includes the contents of absolute paths.

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 6]
```diff
@@ -183,6 +191,38 @@ class ScreenshotCommand extends FlutterCommand {
...
+
+    // Conditional for validating filename
+    if (outputFile.basename.isEmpty || outputFile.basename == '.') {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/screenshot.dart:218]: what about `..` or `../`?

File: packages/flutter_tools/test/commands.shard/hermetic/screenshot_command_test.dart [hunk 7]
```diff
@@ -26,42 +27,89 @@ void main() {
...
     testUsingContext('device screenshots cannot provided Observatory', () async {
-      await expectLater(() => createTestCommandRunner(ScreenshotCommand())
+      await expectLater(() => createTestCommandRunner(ScreenshotCommand(fs: MemoryFileSystem.test()))
         .run(<String>['screenshot',  '--observatory-url=http://localhost:8181']),
         throwsToolExit(message: 'Observatory URI cannot be provided for screenshot type device'),
       );
     });
   });
+
+  group('Screenshot file validation', () {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/screenshot_command_test.dart:69]: Since all this command does now is check if the file exists, you only need a test case for 1) file exists and 2) file doesn't exist

**Review by jonahwilliams — APPROVED** [review]
LGTM
